### ISSUE TITLE
perf(model): amortize Qwen Metal decode through the whole-token graph

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12717]
For: Mac users serving dense Qwen3.8 through the existing fak-native Metal resident sequence route.

Problem: after P32 prefill promotes resident GDN owners, each P1 decode still enters the per-layer host loop. The existing whole-model projection graph rejects P1, retaining repeated command-buffer completion and intermediate host handoffs across the transformer.

Today: the normal resident owner uses per-layer decode blocks/mixers. Historical exact-M3 profiling selects synchronization-bound work; its wait counter includes GPU execution and is not all removable overhead. Simply enabling existing Q8 grouping was already rejected in #8833/#9230 (+124% wall), so that selector remains unchanged.

Better because: reuse the existing whole-model graph for exactly one continuation token when a compatible resident owner is already active, with one completion boundary and final hidden readback. Preserve the established P32 prefill contract and unsupported-path behavior.

Scope: `internal/model/metal_prefill_hybrid.go`, `internal/model/quant_forward.go`, and one focused Darwin model test. No new kernels, no automatic expansion to unsupported routes, no change to Q8 grouping, MTP, public APIs, or provider selection.

Working spine: P32 native prefill -> resident-state handoff -> P1 graph -> final-normalized hidden -> existing output head. Handle packed embedding row gathering, cache position/KV ownership, diagnostic opt-outs, and accepted-error failure without replay.

Witness: first capture a failing P32-to-P1 lifecycle/parity regression; then prove numerical agreement and greedy token parity, one command buffer/completion wait and zero intermediate graph readbacks on physical Apple Silicon. Follow with exact-artifact fak-native control/candidate measurements bound to source, configuration, load, memory, and quality. Retain or reject based on observed net benefit. A synthetic witness alone cannot establish exact-model performance or default promotion.

Centrality: Core, portfolio tier 2; enables responsive local agent serving.
P1: preserved - resident recurrent and attention state and cache positions retain their established ownership semantics.
P2: advanced - measure the full token path including graph creation, uploads, synchronization, head, and state bookkeeping before promotion.
P3: preserved - selection stays bounded to the existing supported owner route; pre-submit declines remain safe and post-submit errors fail closed.
P4: advanced - preserve executed route receipts and report actual physical hardware evidence.

Coordinates with: #12587
Coordinates with: #10193
No start blocker. Default promotion and long-context performance claims require separate exact-envelope qualification.

Duplicate search: open and closed Metal whole-token/decode and Q8-grouping issues reviewed. #8324/#9230/#9456 are closed historical parent/sequence work; current P1 admission is still absent.

Done condition: focused fix, independent review, physical parity plus net-performance witness, explicit-path verified commit and safe push. If a candidate fails net benefit, keep the rejection evidence and iterate on the observed cause.

## Dispatch contract repair (2026-09-12)

Centrality: Core
P1: preserved - resident recurrent state and cache positions retain their established ownership semantics.
P2: advanced - the candidate measures the complete token path before any promotion claim.
P3: preserved - selection stays bounded to the existing compatible resident route with safe declines.
P4: advanced - route counters and physical parity make the executed path observable.

```routing
lane: model
paths: ["internal/model/metal_prefill_hybrid.go", "internal/model/metal_prefill_hybrid_test.go", "internal/model/quant_forward.go"]
expected_steps: 5
```

## Parent context
#12717; Coordinates with: #12587; Coordinates with: #10193

## Current state
After P32 prefill establishes resident GDN owners, P1 decode still takes the per-layer host loop because the whole-model projection graph rejects a one-token continuation.

## Why this is next
The bounded P32-to-P1 lifecycle seam is the next native Metal latency lever after prior Q8 grouping was measured and rejected.

## Working spine
Capture P32 resident-state handoff into P1, admit the compatible whole-token graph, preserve cache/GDN ownership, and return only final normalized hidden state to the existing output head.

## Scope
The first executable slice owns the full issue outcome in the three routed files: lifecycle/parity regression, compatible P1 selection, one completion boundary, no intermediate readback, and fail-closed post-submit behavior.

## Gold-plating boundary
No new kernel, Q8 grouping change, MTP work, public API, unsupported-route expansion, or default performance claim without exact-envelope evidence.

## Done condition
The compatible resident route achieves numerical and greedy-token parity with one command-buffer completion and zero intermediate graph readbacks; unsupported routes remain unchanged.

## Witness
`go test ./internal/model -run '^TestMetalQwen35P32ToP1WholeTokenGraph$' -count=1`

## Acceptance gate
The focused lifecycle/parity test and owned-path validation pass; closure additionally requires physical Apple Silicon counters and exact-artifact net-performance evidence.

## Lane
model

## Closure binding
The resolving explicit-path commit cites #12717 and carries `(fak model)`; #12587/#10193 are nonblocking coordination and exact-envelope qualification gates promotion claims.

**Comment by anthony-chaudhary** [comment]
2026-09-14 live Mac refresh — the whole-token decode graph for this issue is **authored and measured, but not landed**.

Branch (local, unpushed, in the `fak` repo): `fak/w1-whole-token-decode` @ `a94371d13`, four commits on base `6ff7735e9` (which IS the current `origin/main` tip — rebase-free):

- `4cc4eb164` whole-token Qwen3.8 decode graph with P=1 GEMV + per-shape buffer pool
- `4e3ba8996` split-KV flash decoding for P=1 graph attention
- `8aac1d7c5` gather Qwen35 prefill embeddings through the packed store
- `a94371d13` uncap Qwen ordered-panel attention with online softmax

Measured on Apple M3 Pro, Q4_K_M, `fak serve --metal --native-context-tokens 20480` (SW-VERIFIED, host under concurrent fleet load):

| route | decode tok/s | GPU util |
|---|---|---|
| plain resident-Q4K (current main path) | 0.4–0.91 | 24% |
| whole-token graph (this branch) | 4.2–4.6 | 96–97% |

So ~5–10x decode and GPU util from 24% -> 96%. Output numerically coherent (counting prompt).

**Two gaps remain before this closes:**

1. **Land:** none of the four commits is an ancestor of `fak` origin/main (`git merge-base --is-ancestor 4cc4eb164 origin/main` -> false). Land the branch (base = main tip, no rebase needed).
2. **Wire:** the graph is gated by `FAK_QWEN35_WHOLE_TOKEN_DECODE=1` (`internal/model/qwen35_decode_block.go:63`, `qwen35_prefill_q4k.go:86`) and no `cmd/fak` path sets it, so `fak up` never gets the fast path out of the box. Admit it behind auto/profiled admission.

Related unpushed siblings sharing the same base: `fak/w2-q8-kv` (Q8_0 KV realization + `--kv-precision`) and `fak/w3-serving-correctness` (healthz 503 while warming + drop the 120s SSE total timeout).

Still open beyond this fix: prefill is 4.9–7.9 tok/s (target >30) and is sync/dispatch-bound — a separate, larger batched-prefill effort.

**Comment by anthony-chaudhary** [comment]
**Verdict: the whole-token Qwen3.8 P1 decode route is landed on `origin/main`, default-on, and witnessed on physical Apple Silicon — closing.** The 2026-09-14 status refresh above (branch authored, not landed; gated off by default) is stale on both counts: all six commits are ancestors of `origin/main`, and the route defaults ON (opt-out `FAK_QWEN35_WHOLE_TOKEN_DECODE=0`, `internal/model/qwen35_decode_block.go:66`).

| commit | what it did | landing |
|---|---|---|
| `4cc4eb164` | whole-token Qwen3.8 decode graph, P=1 GEMV + per-shape buffer pool | on `origin/main` |
| `4e3ba8996` | split-KV flash decoding for P=1 graph attention | on `origin/main` |
| `8aac1d7c5` | gather Qwen35 metal prefill embeddings via packed store | on `origin/main` |
| `a94371d13` | uncap Qwen ordered-panel attention with online softmax | on `origin/main` |
| `684ccb9eb` | default the whole-token decode route on | on `origin/main` |
| `e32b6ea44` | report the auto-admitted route truthfully | on `origin/main` |

**Witness run this host** (physical Apple M3 Pro, darwin/arm64, macOS 26.6.2, Go 1.26.6) — both green:

```
go test ./internal/model -run 'TestQwen35WholeTokenDecodeRecordsExecutedForwardReceipt|TestQwen35DecodeCallerOwnedGraphParityAndSingleSubmission|TestExactQwen38SessionSelectsWholeDecodeBlockAndHooksOnce|TestExactQwen38DecodeBlockPostSubmitFailureIsTerminal' -count=1
# -> ok; 4/4 PASS

go test ./internal/model -run 'TestMetalQwen35BackendNilP32WholeSequenceSingleFenceAndDecodeOwner|TestMetalQwen35GDNProductionPrefillContinuityAndDecodeHandoff|TestMetalQwen35BackendNilP32PostSubmitFailureIsTerminal|TestQwen35MetalDecodeBlockFourStepsParityAccountingAndGreedyToken' -count=1
# -> ok 8.750s; 4/4 PASS
#    prefill logits cosine=1.000000000 maxRel=3.29216e-06 greedy=410
#    first decode logits cosine=1.000000000 maxRel=1.60961e-06 greedy=59
```

These cover the Done condition: P32 resident-state handoff into P1, one command-buffer completion boundary (`CommandBuffers=1, TerminalWaits=1`), zero intermediate readbacks (`IntermediateReadbacks=0`), greedy-token parity, and fail-closed post-submit without replay.

**Honest disclosure on the declared witness name.** The routing block declared witness `TestMetalQwen35P32ToP1WholeTokenGraph`. That test was written against an earlier worker draft that extended `Qwen35MetalForwardSequence` in place to accept `len(ids)==1`; the draft was archived to an orphan ref (`refs/fak/orphaned/fleet-model-468eca59afbe-*`, `02471b9b3`) and never landed. The landed route instead added a dedicated P=1 decoder, `Qwen35MetalDecodeToken` (`internal/model/metal_prefill_hybrid.go:910`), admitted by `tryQwen35MetalDecodeWholeToken` (`internal/model/qwen35_decode_block.go:61`). `Qwen35MetalForwardSequence` intentionally still requires 32 tokens. The archived draft would have shadowed the shipped P=1 GEMV route via a second admission branch in `tokenHiddenQ`, so it was deliberately not landed. The real landed witnesses are the eight tests above.

**Decode number provenance.** The 0.4–0.9 → 4.2–4.6 tok/s and GPU util 24% → 96% figures are the benchmark measurement captured at the landing commits (M3 Pro, Q4_K_M, host under concurrent fleet load), not a re-measurement in this run.

**Durable evidence:** `docs/_witnesses/issue-12717-whole-token-p1/README.md` (landed on `origin/main`).

**Nonblocking remainder — closure does not overclaim prefill performance.** Accelerated/batched Qwen prefill (current 4.9–7.9 tok/s, target >30) remains a separate, larger effort tracked by #11036 under epic #12459. #12587 and #10193 remain nonblocking coordination. No promotion claim beyond the evidenced resident route.

**Comment by anthony-chaudhary** [comment]
Closing: the whole-token P1 decode route is landed on `origin/main` (e32b6ea44 atop 684ccb9eb) and witnessed green on physical Apple M3 Pro; durable evidence at docs/_witnesses/issue-12717-whole-token-p1/README.md. See the evidence comment above. Reopen if the resident route regresses or the declared scope is not met.
